### PR TITLE
feat(release): per-package graduation via graduate:<package> label

### PR DESCRIPTION
**Pull request by goosewobbler**
Closes #486.

## What

Graduating a prerelease package to stable is a per-package decision. Until now the only lever was the global `release:graduate`, which graduates the whole PR. This adds a per-package `graduate:<package>` label so a maintainer can promote one prerelease unit to stable while the rest keep advancing their `-next` line.

## How

- **Label:** `graduate:<package>` (prefix configurable via `ci.labels.graduatePackagePrefix`, default `graduate:`). ReleaseKit seeds one label per currently-prerelease package each standing-PR update so it shows up in GitHub's label picker.
- **Version engine:** new `graduate` run option → scoped `stableOnly` + `graduateScope`. `applyOverrideScope` clears `stableOnly` for out-of-scope packages (they advance along their line); a graduated package drops its prerelease suffix to the stable base.
- **Group atomicity:** graduating any member of a `fixed`/`linked` group graduates the whole group — enforced in the group strategy's `computeGroup`, which also fixes a latent over-bump where global `release:graduate` + groups mis-derived the bump.
- **Manifest:** new optional `graduated` field records what graduated (survives re-runs); `graduate:*` labels join `overrideLabels` so they're covered by the publish staleness guard and the unauthorized-label reconcile.
- `release:graduate` still graduates every queued prerelease.

Out of scope: no changes to the standing-PR selection-region rendering (#487).

## Tests

Unit/integration: single-package graduation in a mixed standing PR (one goes stable + correct channel, others stay prerelease), manifest round-trip, override-label/staleness wiring, fixed & linked group atomicity, label helpers, and the whole-batch-wins precedence.

`pnpm turbo run lint typecheck test` — all 32 tasks green; `schema:check` clean.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details open><summary><h3>Greptile Summary</h3></summary>

This PR adds per-package graduation for prerelease packages in standing-PR mode. A maintainer can now apply a `graduate:<package>` label to promote one package (and, atomically, any `fixed`/`linked` group it belongs to) to its stable base version while other prerelease packages keep advancing their line.

- **Label & config**: new `graduate:<package>` label shape with a configurable `ci.labels.graduatePackagePrefix` (default `"graduate:"`); ReleaseKit seeds one label per currently-prerelease package on each update so the GitHub label picker stays up to date.
- **Version engine**: `VersionRunOptions.graduate` sets `stableOnly + graduateScope` in the engine; `applyOverrideScope` is extended so `graduateScope` authoritatively controls `stableOnly` per-package, fixing a latent bug where a graduate target that was also a transitive prerequisite (outside `overrideScope`) could silently lose its graduation.
- **Group atomicity**: `computeGroup` grows a graduation path that fires when any member is in `graduateScope` and the group's max baseline is prerelease; the whole group drops to the stable base in one shot, and `graduatedPackageNames` widens the manifest's new `graduated` field to every releasing group member.
</details>

<h3>Confidence Score: 5/5</h3>

Safe to merge. The per-package graduation logic is correctly isolated from the existing global graduation path and all shared edge cases are covered by unit tests.

The two scope-interaction invariants that matter most — re-asserting stableOnly for a graduate target that sits outside overrideScope, and gating the group graduation path on semver.prerelease(maxBaseline) — are both correctly implemented and covered by dedicated test suites. The never-regress guard inside the group graduation path is safe because semver places prerelease versions below their stable counterpart. No logic gaps found beyond the stale-label UX limitation already flagged in prior review threads.

No files require special attention. The stale graduate label cleanup after graduation is a known open UX issue tracked in prior threads and does not affect correctness.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| packages/version/src/core/versionCalculator.ts | applyOverrideScope extended to handle graduateScope: re-asserts stableOnly for packages inside the scope, clearing it for those outside. Correctly handles the prerequisite-outside-overrideScope edge case via sequential application of both scope checks. |
| packages/version/src/core/groupStrategy.ts | New isGroupGraduating helper and graduation path in computeGroup. Group atomicity is enforced: when any member is in graduateScope and the group's max baseline is prerelease, all releasing members adopt the stable base. Never-regress guard correctly keeps prerelease ownNext values from superseding the stable groupVersion. |
| packages/version/src/core/versionEngine.ts | Folds runOptions.graduate into stableOnly + graduateScope when the global stable flag is not set. Global stable correctly wins over per-package graduate. |
| packages/release/src/standing-pr/standing-pr.ts | Core orchestration: resolveStandingPrLabelOverrides parses graduate: labels with correct stable-wins precedence and conflict detection; prereleasePackageNames seeds per-package labels for the next picker update; graduatedPackageNames captures group-atomic graduation for the manifest's new graduated field. |
| packages/release/src/label-utils.ts | Adds graduatePackageLabel, isGraduatePackageLabel, and graduatedPackageFromLabel helpers. graduatePrefix safely falls back to the default when the config omits graduatePackagePrefix, preventing empty-prefix footguns. |
| packages/release/src/label-definitions.ts | deriveLabelDefinitions extended to accept graduatablePackages and seed per-package graduate: labels; callers without prerelease context omit the arg, so labels are minted lazily. |
| packages/version/src/types.ts | Adds graduate to VersionRunOptions and graduateScope to Config, both documented as runtime-only overrides that are always set together via the engine. |
| packages/config/src/schema.ts | graduatePackagePrefix added to CILabelsConfigSchema with default "graduate:" and matching default in the CIConfigSchema.labels default object. |

</details>

<details><summary><h3>Flowchart</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Standing PR update triggered] --> B{Parse PR labels}
    B --> C{release:graduate present?}
    C -- Yes --> D[stable = true global graduation]
    C -- No --> E{graduate pkg labels present?}
    E -- Yes --> F[graduate = pkg names from labels]
    E -- No --> G[No graduation]
    D --> H{channel:prerelease also set?}
    F --> H
    H -- Yes --> I[CONFLICT pending status drop graduation]
    H -- No --> J[buildBaseReleaseOptions]
    J --> K[VersionEngine run]
    K --> L{graduate set and not stable?}
    L -- Yes --> M[stableOnly = true graduateScope = pkg names]
    L -- No --> N[stableOnly = true no scope]
    M --> O[applyOverrideScope per package]
    N --> O
    O --> P{package in graduateScope?}
    P -- Yes --> Q[stableOnly = true returns stable base]
    P -- No --> R[stableOnly = undefined advances prerelease]
    P -- empty scope --> S[stableOnly = true for all]
    Q --> T{Fixed or Linked group?}
    R --> T
    T -- Yes --> U[isGroupGraduating check]
    U -- Yes --> V[groupVersion = stable base all members adopt]
    U -- No --> W[Normal group bump]
    T -- No --> X[Per-package stable base]
    V --> Y[action graduated]
    X --> Y
    W --> Z[action bumped]
    Y --> AA[manifest.graduated recorded]
    AA --> AB[syncLabels seed graduate labels for remaining prerelease packages]
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A[Standing PR update triggered] --> B{Parse PR labels}
    B --> C{release:graduate present?}
    C -- Yes --> D[stable = true global graduation]
    C -- No --> E{graduate pkg labels present?}
    E -- Yes --> F[graduate = pkg names from labels]
    E -- No --> G[No graduation]
    D --> H{channel:prerelease also set?}
    F --> H
    H -- Yes --> I[CONFLICT pending status drop graduation]
    H -- No --> J[buildBaseReleaseOptions]
    J --> K[VersionEngine run]
    K --> L{graduate set and not stable?}
    L -- Yes --> M[stableOnly = true graduateScope = pkg names]
    L -- No --> N[stableOnly = true no scope]
    M --> O[applyOverrideScope per package]
    N --> O
    O --> P{package in graduateScope?}
    P -- Yes --> Q[stableOnly = true returns stable base]
    P -- No --> R[stableOnly = undefined advances prerelease]
    P -- empty scope --> S[stableOnly = true for all]
    Q --> T{Fixed or Linked group?}
    R --> T
    T -- Yes --> U[isGroupGraduating check]
    U -- Yes --> V[groupVersion = stable base all members adopt]
    U -- No --> W[Normal group bump]
    T -- No --> X[Per-package stable base]
    V --> Y[action graduated]
    X --> Y
    W --> Z[action bumped]
    Y --> AA[manifest.graduated recorded]
    AA --> AB[syncLabels seed graduate labels for remaining prerelease packages]
```

</a>
</details>

<sub>Reviews (2): Last reviewed commit: ["fix(release): harden per-package graduat..."](https://github.com/goosewobbler/releasekit/commit/0fa6f2452afed8406df22a02dbce1e4f4836c1c2) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=40669578)</sub>

### CODE REVIEW

File: packages/version/src/core/versionCalculator.ts [hunk 1]
```diff
@@ -61,19 +61,28 @@ export function applyOverrideScope(
...
+  // Per-package graduation (#486): graduate only the packages in `graduateScope`; every other
+  // prerelease keeps its line. Gate ONLY the `stableOnly` graduation here — unlike `overrideScope`,
+  // a graduate is bump-less, so leave any bump/prerelease the run also carries untouched. An empty
+  // `graduateScope` with `stableOnly` set means "graduate everything" (the global release:graduate
+  // path), so this clause is a no-op then.
+  if (graduateScope?.length && options.name && !shouldMatchPackageTargets(options.name, graduateScope)) {
+    scoped = { ...scoped, config: { ...scoped.config, stableOnly: undefined } };
+  }
+  return scoped;
 }
```
[the remark this answers is not in the document]
Comment (goosewobbler) [reply packages/version/src/core/versionCalculator.ts:92]: Fixed in 0fa6f245. `graduateScope` membership is now authoritative for `stableOnly` in `applyOverrideScope`: a package inside `graduateScope` re-asserts `stableOnly = true` even when the `overrideScope` clause stripped it (the transitive-prerequisite-outside-overrideScope case). Only `stableOnly` is re-asserted — the bump/`isPrerelease` cleared by the overrideScope branch stay cleared (graduation is bump-less + commit-driven). Safe because the engine only ever sets `graduateScope` together with `stableOnly: true`, so re-asserting can't fabricate an unrequested graduation. Added combo tests to applyOverrideScope.spec.ts.

File: packages/release/src/standing-pr/standing-pr.ts [hunk 2]
```diff
@@ -583,6 +640,19 @@ function resolveStandingPrLabelOverrides(prLabels: string[], ciConfig: CIConfig
...
+  // a per-package subset would only narrow it; drop the per-package set then. `channel:prerelease`
+  // forces a prerelease line, which directly contradicts graduating, so flag it as a conflict.
+  const graduate = stable
+    ? []
+    : prLabels.map((l) => graduatedPackageFromLabel(l, labels)).filter((p): p is string => p !== undefined);
+  if (graduate.length > 0 && hasPrerelease) {
+    conflicts.push(
+      `Conflicting release-type labels on standing PR (per-package graduate and ${labels.prerelease}) — a package can't graduate and stay prerelease`,
+    );
+  }
```
[the remark this answers is not in the document]
Comment (goosewobbler) [reply packages/release/src/standing-pr/standing-pr.ts:660]: Fixed in 0fa6f245 by naming the offending `graduate:<package>` label(s) in the conflict status (front-loaded so they survive the 140-char status-check truncation). I deliberately did NOT auto-remove the label: the standing PR recomputes version state *from the labels* every run, and `manifest.graduated` is a provenance record, not a recompute input. Removing the label (which also fires an `unlabeled` re-run) would make the next recompute see the still-prerelease package on `main` with no graduate label and revert the graduation to a `-next` bump before the PR ever merges. Naming the label resolves the opaque-conflict UX without that risk.
